File: code/198.go
/*
 * @Author: lmio
 * @Date: 2023-02-11 21:09:24
 * @LastEditTime: 2023-04-07 14:35:28
 * @FilePath: /leetcode/code/198.go
 * @Description:198.打家劫舍
 */
package code

import "leetcode/utils"

func RobI(nums []int) int {
	length := len(nums)
	if length == 1 {
		return nums[0]
	}else if length == 2 {
		return utils.Max(nums[0], nums[1])
	}
	dp := []int{}
	dp = append(dp, nums[0], utils.Max(nums[0], nums[1]))
	
	for i := 2; i < length; i++ {
		dp = append(dp, utils.Max(dp[i-1], dp[i-2] + nums[i]))  
	}
	return dp[length-1]
}

File: code/198.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-09-16 18:22:12
LastEditors: lmio 2091319361@qq.com
Description: 198. 打家劫舍
'''

from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        dp = [0]*n
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, n):
            dp[i] = max(dp[i-1], nums[i]+dp[i-2])
        return dp[-1]